src/base/misc/disks.c

```diff
@@ -929,163 +916,100 @@ static void image_setup(struct disk *dp)
- *       of the drive. The physical h/s/c start and end are calculated and
- *       put in the dp->part_info.number'th entry in the part table.
- */
-#ifdef __linux__
-static void set_part_ent(struct disk *dp, unsigned char *tmp_mbr)
+static void partition_setup(struct disk *dp)
 {
-  long	length;		/* partition length in sectors		*/
-  long	end;		/* last sector number offset		*/
-  unsigned char	*p;	/* ptr to part table entry to create	*/
+  struct on_disk_mbr *mbr;
+  unsigned char sect[512];
 
-  if (!dp->part_image) {
-    if (ioctl(dp->fdesc, BLKGETSIZE, &length)) {
-      error("calling ioctl BLKGETSIZE for PARTITION %s\n", dp->dev_name);
-      leavedos(22);
-    }
+  d_printf("PARTITION SETUP for %s\n", dp->dev_name);
+
+  if (dp->floppy) {
+    return;
+  }
+
+  dp->part_info.number = 1;
+  dp->part_info.mbr_size = sizeof(*mbr);
+  dp->part_info.mbr = malloc(dp->part_info.mbr_size);
+  mbr = (struct on_disk_mbr *)dp->part_info.mbr;
+
+  memset(mbr, 0, sizeof(*mbr));
+  memcpy(mbr->code, mbr_boot_code, sizeof(mbr_boot_code));
+  mbr->partition[0] = build_pi(dp);
+  mbr->signature = MBR_SIG;
+
+  lseek(dp->fdesc, 0, SEEK_SET);
+  if (RPT_SYSCALL(read(dp->fdesc, sect, sizeof(sect))) != sizeof(sect)) {
+    d_printf("  bpb could not be read\n");
   } else {
-    struct stat sb;
-    if (fstat(dp->fdesc, &sb)) {
-      error("calling ioctl BLKGETSIZE for PARTITION %s\n", dp->dev_name);
-      leavedos(22);
-    }
-    length = sb.st_size;
-  }
-#define SECPERCYL	(dp->heads * dp->sectors)
-#define CYL(s)		((s)/SECPERCYL)			/* 0-based */
-#define HEAD(s)		(((s)%SECPERCYL)/dp->sectors)	/* 0-based */
-#define SECT(s)		(((s)%dp->sectors)+1)		/* 1-based */
-
-  d_printf("SET_PART_ENT: making part table entry for device %s,\n",
-	dp->dev_name);
-  d_printf("Calculated physical start: head=%4ld sect=%4ld cyl=%4ld,\n",
-	HEAD(dp->start), SECT(dp->start), CYL(dp->start));
-  end = dp->start+length-1;
-  d_printf("Calculated physical end:   head=%4ld sect=%4ld cyl=%4ld.\n",
-	HEAD(end), SECT(end), CYL(end));
-
-  /* get address of where to put new part table entry */
-  p = tmp_mbr + PART_INFO_START + (PART_INFO_LEN * (dp->part_info.number-1));
-
-  /* initialize with LBA values */
-  p[1] = p[5] = 254;
-  p[2] = p[6] = 255;
-  p[3] = p[7] = 255;
-  p[4] = 0xe;
-
-  p[0] = PART_BOOT;						/* bootable  */
-  if (CYL(dp->start) <= 1023) {
-    p[1] = HEAD(dp->start);					/* beg head  */
-    p[2] = SECT(dp->start) | ((CYL(dp->start) >> 2) & 0xC0);	/* beg sect  */
-    p[3] = CYL(dp->start) & 0xFF;				/* beg cyl   */
-  }
-  if (CYL(end) <= 1023) {
-    p[4] = (length < 32*1024*1024/SECTOR_SIZE)? 0x04 : 0x06;	/* dos sysid */
-    p[5] = HEAD(end);						/* end head  */
-    p[6] = SECT(end) | ((CYL(end) >> 2) & 0xC0);		/* end sect  */
-    p[7] = CYL(end) & 0xFF;					/* end cyl   */
-  }
-  *((uint32_t *)(p+8)) = dp->start;				/* pre sects */
-  *((uint32_t *)(p+12)) = length;				/* len sects */
+    struct on_disk_vbr *vbr = sect;
```

Comment:
Even w/o any cast?
You did that with a cast in
another place.

Reply:
Union os a byte array and vbr
may be an adequate solution
if the size of the struct is smaller
than a sector, or if you need to
have it as an array too.

Reply:
> Even w/o any cast?

That was sloppy!

I have to go out now, so will fix later, thanks!

Reply:
I added a new bpb7 structure type and added a union into the vbr to avoid the pointer troubles. If you are okay with this I'll squash those commits into the earlier ones.